File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? {
    // check all values in our list against our predicate
    strings.forEach { string -> if (predicate(string)) { return string.lowercase() } }
    // if nothing passed the predicate
    return null
}

// create a say class where the default phrase is the empty string
class Say(val toSay: String = "") {
    // set our phrase property to what we want to say
    val phrase: String = toSay

    // if we want to add a word to our phrase
    fun and(toSay: String): Say {
        return Say(phrase + " " + toSay)
    }
}
// say with a parameter
fun say(toSay: String = ""): Say {
    return Say(toSay)
}

fun meaningfulLineCount(filePath: String): Long {
    // create a counter for our meaningful lines (must be a long to match return type)
    var goodLines: Long = 0L

    // use BufferedReader to go through each line of the file
    BufferedReader(FileReader(filePath)).use { reader ->
        // read the first line of the file (it will be null if there are no lines in the file)
        var line: String? = reader.readLine()
        // loop until we reach the end of the file
        while (line != null) {
            // if the line is meaningful
            if (line.trim().isNotEmpty() && line.trim().substring(0, 1) != "#") {
                // count it
                goodLines++
            }
            // read the next line (it will be null if we are currently on the last line)
            line = reader.readLine()
        }
    }
    // return our count
    return goodLines
}

// Write your Quaternion data class here

// Write your Binary Search Tree interface and implementing classes here

File: swift/exercises.swift
import Foundation

struct NegativeAmountError: Error {}
struct NoSuchFileError: Error {}

func change(_ amount: Int) -> Result<[Int:Int], NegativeAmountError> {
    if amount < 0 {
        return .failure(NegativeAmountError())
    }
    var (counts, remaining) = ([Int:Int](), amount)
    for denomination in [25, 10, 5, 1] {
        (counts[denomination], remaining) = 
            remaining.quotientAndRemainder(dividingBy: denomination)
    }
    return .success(counts)
}

func firstThenLowerCase(of strings: [String], satisfying predicate: (String) -> Bool) -> String? {
    // for each string in the list of strings
    for string in strings {
        // if it passes the predicate
        if predicate(string) {
            // return the lower case string
            return string.lowercased()
        } 
    }
    // if no string passed the predicate, return nil
    return nil
}

// object for say
struct Say {
    // store our phrase
    let phrase: String

    // set the phrase to the given string
    init(_ toSay: String) {
        self.phrase = toSay
    }

    // create a new Say with the additional word
    func and(_ toSay: String) -> Say {
        return Say(self.phrase + " " + toSay)
    }
}
// function to create a say object with a default of the empty string
func say(_ toSay: String = "") -> Say {
    return Say(toSay)
}

// Handle errors while reading file
enum FileReadError: Error {
    case noSuchFile
}
// meaninful line counter
func meaningfulLineCount(_ filePath: String) async -> Result<Int, FileReadError> {
    // store our meaningfulLines
    var goodLines = 0

    // make sure we are able to read the file
    do {
        // get the contents of the file
        let fileContents = try String(contentsOfFile: filePath, encoding: .utf8)

        // check each line in the file
        for line in fileContents.split(separator: "\n") {
            // trim whitespace off the line
            let trimmedLine = line.trimmingCharacters(in: .whitespacesAndNewlines)

            // check if the line is meaningful (not empty and no leading #)
            if (trimmedLine != "" && trimmedLine[trimmedLine.startIndex] != "#") {
                goodLines += 1
            }
        }
    } catch {
        // if we got an error trying to read the file, return a failure Result
        return .failure(.noSuchFile)
    }

    // if everything went right, return the total number of meaningful lines
    return .success(goodLines)
}

// Write your Quaternion struct here

// Write your Binary Search Tree enum here
enum BinarySearchTree {
    // list the possible cases
    case empty
    indirect case node(BinarySearchTree, Int, BinarySearchTree)

    // return the number of nodes in our tree
    var size: Int {
        switch self{
            // if it's empty, return 0
            case .empty: return 0
            // if it isn't empty, get the size of its children and add 1 for the size of the current node
            case let .node(left, _, right): return left.size + 1 + right.size
        }
    }
}
